### PR TITLE
[AMD] amd/deepseek_v4 integration 28/N use aiter greedy_sample for all-greedy sampling on ROCm

### PR DESCRIPTION
**Pull request by yichiche**
Update amd/deepseek_v4 integration branch

Following PRs have large set of conflict, we use this PR and upstream amd/deepseek_v4 branch to integrate in parallel.
https://github.com/sgl-project/sglang/pull/23600
https://github.com/sgl-project/sglang/pull/23608

co-author: @hubertlu-tw [17095](https://github.com/sgl-project/sglang/pull/17095)
## Motivation

The Sampler's all-greedy path currently uses `torch.argmax`, which dispatches to a generic `reduce_kernel` (ArgMaxOps). On ROCm/MI300, this kernel takes ~40.2 us per call and accounts for 7.4% of total kernel time during decode.

aiter provides a specialized `greedy_sample` kernel optimized for the argmax-over-vocabulary use case. This PR replaces `torch.argmax` with `aiter.greedy_sample` when running on HIP with `SGLANG_USE_AITER` enabled.

## Modifications

- **`python/sglang/srt/layers/sampler.py`**:
  - Import `is_hip` and conditionally import `aiter.greedy_sample` when `SGLANG_USE_AITER` is set on HIP.
  - In the all-greedy sampling path, use `aiter.greedy_sample` (pre-allocating an int32 output tensor) instead of `torch.argmax`.
  - Falls back to `torch.argmax` on non-HIP backends or when aiter is not enabled.

- **`test/registered/ops/test_aiter_greedy_sample_amd.py`** :
  - **Kernel-level tests** (`TestAiterGreedySample`): validates `aiter.greedy_sample` produces identical results to `torch.argmax` across realistic LLM vocab sizes (32k–152k), various batch sizes (1–256), and edge cases (tied values, negative logits, extreme values).
  - **Sampler integration tests** (`TestAiterGreedyIntegration`): exercises `Sampler.forward()` through the actual `_use_aiter` code path via `mock.patch`, covering:
    - aiter vs argmax token-ID equivalence through the full sampler
    - fallback to `torch.argmax` when `_use_aiter=False`
    - `return_logprob=True` path (logprob gather uses `batch_next_token_ids`)
    - output dtype contract (aiter returns int32, argmax returns int64)
    - various batch sizes through the sampler
  - Registered in AMD CI: `register_amd_ci(est_time=60, suite="stage-b-test-1-gpu-small-amd")`
  
## Accuracy Tests

| Metric | Value |
|---|---|
| Accuracy | 0.905 |
| Invalid | 0.001 |
| Latency | 240.024 s |
| Output throughput | 536.122 token/s |

## Benchmarking and Profiling

Sampler greedy kernel replacement (per call, ~2,050 calls during decode):

| Kernel | Before (us) | After (us) |
|---|---|---|
| `reduce_kernel` (ArgMaxOps) | 40.2 | — |
| `aiter::greedy_sample_kernel` | — | 10.3 |
| **Savings** | | **29.9 us/call (74.4% reduction)** |
| **% of total kernel time** | 7.4% | 2.2% |

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: **Missing `run-ci` label** — add it to run CI tests.
Latest PR Test (Extra): :x: **Blocked** — `run-ci` is required first.

### CODE REVIEW

File: python/sglang/srt/layers/sampler.py [hunk 1]
```diff
@@ -109,8 +114,13 @@ def forward(
         logits = self._preprocess_logits(logits, sampling_info)
 
         if sampling_info.is_all_greedy:
-            # Use torch.argmax if all requests use greedy sampling
-            batch_next_token_ids = torch.argmax(logits, -1)
+            if _use_aiter:
```
Comment (hubertlu-tw) [inline python/sglang/srt/layers/sampler.py:117]: Hi @yichiche, do you mind adding an unit test script for this greedy sampling kernel like https://github.com/sgl-project/sglang/pull/17095 in case aiter in the future brings in any breakages in the future? Thanks.

**Review by hubertlu-tw — CHANGES_REQUESTED** [review]

**Comment by HaiShaw** [comment]
@yichiche address comment?

File: python/sglang/srt/layers/sampler.py (hunk 1 above)
Comment (yichiche) [reply python/sglang/srt/layers/sampler.py:117]: Done — added test/registered/ops/test_aiter_greedy_sample_amd.py in 75d5785.
1. aiter.greedy_sample produces identical results to torch.argmax (kernel level)
2. Sampler.forward() correctly dispatches to aiter when _use_aiter=True
3. The fallback to torch.argmax works when _use_aiter=False
4. return_logprob path works with the aiter greedy branch

Command: `pytest ./test/registered/ops/test_aiter_greedy_sample_amd.py`

**Comment by yichiche** [comment]
> @yichiche address comment?

Yes, problem solved. I added a unit test to address the points mentioned.

**Review by hubertlu-tw — APPROVED** [review]
LGTM

**Review by HaiShaw — APPROVED** [review]
